rpcserver.go

```diff
@@ -7768,81 +7759,62 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 	}
 
-	var acceptedQuote *rfqrpc.PeerAcceptedBuyQuote
-	switch r := resp.Response.(type) {
-	case *rfqrpc.AddAssetBuyOrderResponse_AcceptedQuote:
-		acceptedQuote = r.AcceptedQuote
+	acquiredQuotes := make([]quoteWithInfo, 0)
 
-	case *rfqrpc.AddAssetBuyOrderResponse_InvalidQuote:
-		return nil, fmt.Errorf("peer %v sent back an invalid quote, "+
-			"status: %v", r.InvalidQuote.Peer,
-			r.InvalidQuote.Status.String())
+	// TODO(george): should we limit the max number of acquired quotes?
+	// Tapd nodes with hundreds or more of chans might face a bottleneck
+	// here. Could just acquire quotes that reach a certain threshold.
+	for peer, channels := range chanMap {
+		quote, err := r.AcquireBuyOrder(
+			ctx, &rpcSpecifier, req.AssetAmount, expiryTimestamp,
+			&peer,
+		)
+		if err != nil {
+			// TODO(george): should handle negotiation failures
+			// gracefully.
+			return nil, err
+		}
 
-	case *rfqrpc.AddAssetBuyOrderResponse_RejectedQuote:
-		return nil, fmt.Errorf("peer %v rejected the quote, code: %v, "+
-			"error message: %v", r.RejectedQuote.Peer,
-			r.RejectedQuote.ErrorCode, r.RejectedQuote.ErrorMessage)
+		rate, err := rfqrpc.UnmarshalFixedPoint(quote.AskAssetRate)
+		if err != nil {
+			return nil, err
+		}
 
-	default:
-		return nil, fmt.Errorf("unexpected response type: %T", r)
+		acquiredQuotes = append(acquiredQuotes, quoteWithInfo{
+			quote: quote,
+			rate:  rate,
+			// Since the channels are sorted, we now the value with
+			// the greatest remote balance is at index 0.
+			channel: channels[0],
+		})
 	}
 
-	// If the invoice is for an asset unit amount smaller than the minimal
-	// transportable amount, we'll return an error, as it wouldn't be
-	// payable by the network.
-	if acceptedQuote.MinTransportableUnits > req.AssetAmount {
-		return nil, fmt.Errorf("cannot create invoice over %d asset "+
-			"units, as the minimal transportable amount is %d "+
-			"units with the current rate of %v units/BTC",
-			req.AssetAmount, acceptedQuote.MinTransportableUnits,
-			acceptedQuote.AskAssetRate)
-	}
+	// Let's sort the ask rate of the quotes in ascending order.
+	sort.Slice(acquiredQuotes, func(i, j int) bool {
+		return acquiredQuotes[i].rate.ToUint64() <
+			acquiredQuotes[j].rate.ToUint64()
+	})
 
-	// Now that we have the accepted quote, we know the amount in Satoshi
-	// that we need to pay. We can now update the invoice with this amount.
-	//
-	// First, un-marshall the ask asset rate from the accepted quote.
-	askAssetRate, err := rfqrpc.UnmarshalFixedPoint(
-		acceptedQuote.AskAssetRate,
-	)
-	if err != nil {
-		return nil, fmt.Errorf("error unmarshalling ask asset rate: %w",
-			err)
-	}
+	// We grab the most expensive rate to use as reference for the total
+	// invoice amount. Since peers have varying prices for the assets, we
+	// pick the most expensive rate in order to allow for any combination of
+	// MPP shards through our set of chosen peers.
+	expensiveRate := acquiredQuotes[0].rate
```

Comment:
Why should we pick the most expensive quote? This is effectively the same as giving the sender an option for the most expensive route in terms of fees.